File: python/day8.py
import math
from collections import defaultdict


def cell2index(y: int, x: int, n_cols: int) -> int:
    return y * n_cols + x


def is_oob(x: int, y: int, n_cols: int, n_rows: int) -> bool:
    return x < 0 or x >= n_cols or y < 0 or y >= n_rows


def add_antinode(antinodes: set[int], y: int, x: int, n_cols: int, n_rows: int) -> None:
    if not is_oob(x, y, n_cols, n_rows):
        antinodes.add(cell2index(y, x, n_cols))


def add_antinodes_line(
    antinodes: set[int], y: int, x: int, dy: int, dx: int, n_cols: int, n_rows: int
) -> None:
    i = 0
    while not is_oob(x + i * dx, y + i * dy, n_cols, n_rows):
        antinodes.add(cell2index(y + i * dy, x + i * dx, n_cols))
        i += 1

    i = 0
    while not is_oob(x + i * dx, y + i * dy, n_cols, n_rows):
        antinodes.add(cell2index(y + i * dy, x + i * dx, n_cols))
        i -= 1


def day8a() -> None:
    rows = []
    with open("../inputs/day8.txt", "r") as file:
        for line in file.readlines():
            rows.append(line.strip())

    antennas = defaultdict(list)
    for i, row in enumerate(rows):
        for j, chr in enumerate(row):
            if chr != ".":
                antennas[chr].append((i, j))

    antinodes = set()
    for ants in antennas.values():
        for i in range(len(ants)):
            for j in range(i + 1, len(ants)):
                dy = ants[i][0] - ants[j][0]
                dx = ants[i][1] - ants[j][1]
                add_antinode(
                    antinodes, ants[i][0] + dy, ants[i][1] + dx, len(rows), len(rows[0])
                )
                add_antinode(
                    antinodes, ants[j][0] - dy, ants[j][1] - dx, len(rows), len(rows[0])
                )
    print(len(antinodes))


def day8b() -> None:
    rows = []
    with open("../inputs/day8.txt", "r") as file:
        for line in file.readlines():
            rows.append(line.strip())

    antennas = defaultdict(list)
    for i, row in enumerate(rows):
        for j, chr in enumerate(row):
            if chr != ".":
                antennas[chr].append((i, j))

    antinodes = set()
    for ants in antennas.values():
        for i in range(len(ants)):
            for j in range(i + 1, len(ants)):
                dy = ants[i][0] - ants[j][0]
                dx = ants[i][1] - ants[j][1]
                gcd = math.gcd(dy, dx)
                dy /= gcd
                dx /= gcd

                add_antinodes_line(
                    antinodes, ants[i][0], ants[i][1], dy, dx, len(rows), len(rows[0])
                )

    print(len(antinodes))


day8b()

File: go/day8/day8.go
package day8

import (
	"fmt"

	"github.com/tadejsv/aoc-2024/utils"
)

func hashCell(y, x, nCols int) int {
	return y*nCols + x
}

func isOOB(y, x, nRows, nCols int) bool {
	return y < 0 || x < 0 || y >= nRows || x >= nCols
}

func addAntiNode(antiNodes map[int]bool, y, x, nRows, nCols int) {
	if !isOOB(y, x, nRows, nCols) {
		antiNodes[hashCell(y, x, nCols)] = true
	}
}

func gcd(x, y int) int {
	if y == 0 {
		return x
	}
	return gcd(y, x%y)
}

func addAntiNodesLine(antiNodes map[int]bool, y, x, dy, dx, nRows, nCols int) {
	for i := 0; isOOB(y+i*dy, x+i*dx, nRows, nCols); i++ {
		antiNodes[hashCell(y+i*dy, x+i*dx, nCols)] = true
	}
	for i := 0; isOOB(y+i*dy, x+i*dx, nRows, nCols); i-- {
		antiNodes[hashCell(y+i*dy, x+i*dx, nCols)] = true
	}
}

func Day8a() {
	lines, err := utils.ReadLines("../inputs/day8.txt")
	if err != nil {
		panic("Could not read lines")
	}

	nRows, nCols := len(lines), len(lines[0])
	antennas := make(map[rune][][2]int)
	for i, line := range lines {
		for j, chr := range line {
			if chr != '.' {
				antennas[chr] = append(antennas[chr], [2]int{i, j})
			}
		}
	}

	antinodes := make(map[int]bool)
	for _, ants := range antennas {
		for i := range ants {
			for j := i + 1; j < len(ants); j++ {
				dy := ants[i][0] - ants[j][0]
				dx := ants[i][1] - ants[j][1]
				addAntiNode(antinodes, ants[i][0]+dy, ants[i][1]+dx, nRows, nCols)
				addAntiNode(antinodes, ants[j][0]-dy, ants[j][1]-dx, nRows, nCols)
			}
		}
	}
	fmt.Printf("%v\n", len(antinodes))
}

func Day8b() {
	lines, err := utils.ReadLines("../inputs/day8.txt")
	if err != nil {
		panic("Could not read lines")
	}

	nRows, nCols := len(lines), len(lines[0])
	antennas := make(map[rune][][2]int)
	for i, line := range lines {
		for j, chr := range line {
			if chr != '.' {
				antennas[chr] = append(antennas[chr], [2]int{i, j})
			}
		}
	}

	antinodes := make(map[int]bool)
	for _, ants := range antennas {
		for i := range ants {
			for j := i + 1; j < len(ants); j++ {
				dy := ants[i][0] - ants[j][0]
				dx := ants[i][1] - ants[j][1]
				gcd := gcd(dy, dx)
				dy, dx = dy/gcd, dx/gcd

				addAntiNodesLine(antinodes, ants[i][0], ants[i][1], dy, dx, nRows, nCols)
			}
		}
	}
	fmt.Printf("%v\n", len(antinodes))
}
